File: backend/app/services/chat_service.py
import logging
from functools import lru_cache
from typing import AsyncIterator, Optional

logger = logging.getLogger(__name__)

from app.core.exceptions import SessionNotFoundException
from app.schemas.base import generate_id
from app.schemas.chat import (
    ChatCompletionRequest,
    ChatCompletionResponse,
    Message,
    Session,
    SessionCreate,
    SessionDetailResponse,
    SessionResponse,
)
from app.schemas.sse import (
    SSEEventType,
    create_content_delta_event,
    create_done_event,
    create_error_event,
    create_message_done_event,
    create_message_start_event,
)
from app.services.agent_service import AgentService
from app.services.llm.factory import get_default_llm_provider
from app.services.memory_service import MemoryService
from app.services.rule_service import RuleService


class ChatService:
    """聊天服务 - 核心业务逻辑"""

    def __init__(self):
        # 内存存储会话（生产环境应使用数据库）
        self._sessions: dict[str, Session] = {}
        # 集成 Agent、Rule、Memory 服务
        self._agent_service = AgentService()
        self._rule_service = RuleService()
        self._memory_service = MemoryService()

    def _build_system_prompt(
        self,
        agent_id: Optional[str] = None,
        rule_ids: Optional[list[str]] = None,
    ) -> str:
        """构建系统提示词"""
        parts = []

        # 添加 Agent 角色设定
        if agent_id:
            try:
                agent = self._agent_service.get_agent(agent_id)
                parts.append(f"你现在是「{agent.name}」。\n{agent.system_prompt}")
            except Exception as e:
                logger.warning(f"Failed to load agent {agent_id}: {e}")

        # 添加规则
        if rule_ids:
            rules_content = []
            for rule_id in rule_ids:
                try:
                    rule = self._rule_service.get_rule(rule_id)
                    if rule.enabled:
                        rules_content.append(rule.content)
                except Exception as e:
                    logger.warning(f"Failed to load rule {rule_id}: {e}")
            if rules_content:
                parts.append("请遵循以下规则：\n" + "\n".join(f"- {r}" for r in rules_content))

        # 添加记忆上下文
        memory_context = self._memory_service.get_memory_context()
        if memory_context:
            parts.append(memory_context)

        return "\n\n".join(parts) if parts else ""

    def create_session(self, request: SessionCreate) -> Session:
        """
        创建新会话

        Args:
            request: 会话创建请求

        Returns:
            新创建的会话
        """
        session = Session(title=request.title or "New Chat")
        self._sessions[session.id] = session
        return session

    def get_session(self, session_id: str) -> Session:
        """
        获取会话

        Args:
            session_id: 会话 ID

        Returns:
            会话对象

        Raises:
            SessionNotFoundException: 会话不存在
        """
        if session_id not in self._sessions:
            raise SessionNotFoundException(session_id)
        return self._sessions[session_id]

    def list_sessions(self, page: int = 1, page_size: int = 20) -> tuple[list[Session], int]:
        """
        列出会话

        Args:
            page: 页码
            page_size: 每页数量

        Returns:
            (会话列表, 总数)
        """
        sessions = sorted(
            self._sessions.values(),
            key=lambda s: s.updated_at,
            reverse=True,
        )
        total = len(sessions)
        start = (page - 1) * page_size
        end = start + page_size
        return sessions[start:end], total

    def delete_session(self, session_id: str) -> bool:
        """
        删除会话

        Args:
            session_id: 会话 ID

        Returns:
            是否删除成功
        """
        if session_id in self._sessions:
            del self._sessions[session_id]
            return True
        return False

    def session_to_response(self, session: Session) -> SessionResponse:
        """将 Session 转换为 SessionResponse"""
        return SessionResponse(
            id=session.id,
            title=session.title,
            created_at=session.created_at,
            updated_at=session.updated_at,
            message_count=len(session.messages),
        )

    def session_to_detail_response(self, session: Session) -> SessionDetailResponse:
        """将 Session 转换为 SessionDetailResponse"""
        return SessionDetailResponse(
            id=session.id,
            title=session.title,
            created_at=session.created_at,
            updated_at=session.updated_at,
            message_count=len(session.messages),
            messages=session.messages,
        )

    async def chat(self, request: ChatCompletionRequest) -> ChatCompletionResponse:
        """
        非流式聊天

        Args:
            request: 聊天请求

        Returns:
            聊天响应
        """
        # 获取或创建会话
        if request.session_id:
            session = self.get_session(request.session_id)
        else:
            session = self.create_session(SessionCreate())

        # 添加用户消息
        user_message = Message(role="user", content=request.message)
        session.add_message(user_message)

        # 构建 LLM 消息
        messages = [{"role": msg.role, "content": msg.content} for msg in session.messages]

        # 构建系统提示词
        system_prompt = self._build_system_prompt(request.agent_id, request.rule_ids)

        # 调用 LLM
        llm = get_default_llm_provider()
        response = await llm.chat(messages, system_prompt=system_prompt)

        # 添加助手消息
        assistant_message = Message(role="assistant", content=response.content)
        session.add_message(assistant_message)

        # 更新会话标题（如果是第一条用户消息）
        if len(session.messages) == 2:  # 用户消息 + 助手消息
            session.title = request.message[:50] + ("..." if len(request.message) > 50 else "")

        # 尝试提取记忆
        try:
            await self._memory_service.extract_memory_from_conversation(
                request.message, response.content, session.id
            )
        except Exception:
            pass  # 记忆提取失败不影响主流程

        return ChatCompletionResponse(
            session_id=session.id,
            message_id=assistant_message.id,
            role="assistant",
            content=response.content,
        )

    async def chat_stream(self, request: ChatCompletionRequest) -> AsyncIterator[str]:
        """
        流式聊天 - 返回 SSE 格式字符串

        Args:
            request: 聊天请求

        Yields:
            SSE 格式的事件字符串
        """
        message_id = generate_id()
        session_id = request.session_id
        sent_finish = False  # 跟踪是否已发送完成事件

        try:
            # 获取或创建会话
            if session_id:
                try:
                    session = self.get_session(session_id)
                except SessionNotFoundException:
                    yield create_error_event(f"Session {session_id} not found", "SESSION_NOT_FOUND")
                    return
            else:
                session = self.create_session(SessionCreate())
                session_id = session.id

            # 添加用户消息
            user_message = Message(role="user", content=request.message)
            session.add_message(user_message)

            # 发送消息开始事件
            yield create_message_start_event(session_id, message_id)

            # 构建 LLM 消息
            messages = [{"role": msg.role, "content": msg.content} for msg in session.messages]

            # 构建系统提示词
            system_prompt = self._build_system_prompt(request.agent_id, request.rule_ids)

            # 调用 LLM 流式 API
            llm = get_default_llm_provider()
            full_content = ""

            async for chunk in llm.chat_stream(messages, system_prompt=system_prompt):
                if chunk.content:
                    full_content += chunk.content
                    yield create_content_delta_event(session_id, message_id, chunk.content)

                if chunk.finish_reason and not sent_finish:
                    sent_finish = True
                    yield create_message_done_event(session_id, message_id, chunk.finish_reason)

            # 如果没有收到 finish_reason，手动发送完成事件
            if not sent_finish:
                yield create_message_done_event(session_id, message_id, "stop")

            # 添加助手消息到会话
            assistant_message = Message(id=message_id, role="assistant", content=full_content)
            session.add_message(assistant_message)

            # 更新会话标题
            if len(session.messages) == 2:
                session.title = request.message[:50] + ("..." if len(request.message) > 50 else "")

            # 尝试提取记忆
            try:
                await self._memory_service.extract_memory_from_conversation(
                    request.message, full_content, session_id
                )
            except Exception:
                pass  # 记忆提取失败不影响主流程

            # 发送会话完成事件
            yield create_done_event(session_id)

        except Exception as e:
            yield create_error_event(str(e), "CHAT_ERROR", session_id)


# 服务依赖注入
@lru_cache
def get_chat_service() -> ChatService:
    """获取 ChatService 实例（单例）"""
    return ChatService()

File: backend/src/modules/chat/chat.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { SessionEntity } from '../../common/entities/session.entity';
import { MessageEntity, MessageRole } from '../../common/entities/message.entity';
import { AgentService } from '../agent/agent.service';
import { RuleService } from '../rule/rule.service';
import { MemoryService } from '../memory/memory.service';
import { LangGraphService } from '../langgraph/langgraph.service';
import { AppConfigService } from '../../config/config.service';
import { CreateCompletionDto } from './dto/create-completion.dto';
import { CreateSessionDto } from './dto/create-session.dto';
import { UpdateSessionDto } from './dto/update-session.dto';

@Injectable()
export class ChatService {
  constructor(
    @InjectRepository(SessionEntity)
    private sessionRepo: Repository<SessionEntity>,
    @InjectRepository(MessageEntity)
    private messageRepo: Repository<MessageEntity>,
    private agentService: AgentService,
    private ruleService: RuleService,
    private memoryService: MemoryService,
    private langGraphService: LangGraphService,
    private configService: AppConfigService,
  ) {}

  async createCompletion(dto: CreateCompletionDto) {
    const { message, session_id, stream, agent_id, rule_ids, web_search } = dto;

    const session = session_id
      ? await this.getSession(session_id)
      : await this.createSession({ title: 'New Chat' });

    // 保存当前使用的 agent_id 到 session
    if (agent_id && session.agent_id !== agent_id) {
      session.agent_id = agent_id;
      await this.sessionRepo.save(session);
    }

    const userMessage = this.messageRepo.create({
      id: uuidv4(),
      role: 'user' as MessageRole,
      content: message,
      session: { id: session.id } as SessionEntity,
    });
    await this.messageRepo.save(userMessage);

    // Reload session with messages
    const sessionWithMessages = await this.sessionRepo.findOne({
      where: { id: session.id },
      relations: ['messages'],
    });

    const systemPrompt = await this.buildSystemPrompt(agent_id, rule_ids, web_search);
    const messages = sessionWithMessages!.messages.map((m) => ({
      role: m.role,
      content: m.content,
    }));

    if (stream) {
      return { session: sessionWithMessages, stream: true, messages, systemPrompt, agent_id };
    }

    const response = await this.langGraphService.chat(messages, systemPrompt, session.id, agent_id);
    const assistantMessage = this.messageRepo.create({
      id: uuidv4(),
      role: 'assistant' as MessageRole,
      content: response.content,
      session: { id: session.id } as SessionEntity,
    });
    await this.messageRepo.save(assistantMessage);

    session.updated_at = new Date();
    await this.sessionRepo.save(session);

    // 异步生成对话标题
    this.generateTitleIfNeeded(session.id, messages, response.content).catch(() => {});

    return { session: sessionWithMessages, assistantMessage, finish_reason: response.finish_reason };
  }

  async *streamCompletion(
    messages: { role: MessageRole; content: string }[],
    systemPrompt: string,
    sessionId: string,
    messageId: string,
    preferredAgent?: string,
  ): AsyncGenerator<{ event: string; data: any }> {
    yield {
      event: 'message_start',
      data: { session_id: sessionId, message_id: messageId, role: 'assistant' },
    };

    let fullContent = '';
    let hasError = false;

    try {
      for await (const event of this.langGraphService.chatStream(messages, systemPrompt, sessionId, preferredAgent)) {
        const base = { session_id: sessionId, message_id: messageId };

        switch (event.type) {
          case 'text':
            fullContent += event.content;
            yield { event: 'content_delta', data: { ...base, content: event.content } };
            break;

          case 'tool_start':
            yield { event: 'tool_call_start', data: { ...base, tool_call_id: event.toolCallId, tool_name: event.toolName } };
            break;

          case 'tool_delta':
            yield { event: 'tool_call_delta', data: { ...base, tool_call_id: event.toolCallId, args_delta: event.argsDelta } };
            break;

          case 'tool_input':
            yield { event: 'tool_call_input', data: { ...base, tool_call_id: event.toolCallId, tool_name: event.toolName, input: event.args } };
            break;

          case 'tool_output':
            yield { event: 'tool_call_output', data: { ...base, tool_call_id: event.toolCallId, output: event.output } };
            break;

          case 'step_start':
            yield { event: 'step_start', data: base };
            break;

          case 'agent_switched':
            yield {
              event: 'agent_switched',
              data: { ...base, from: event.fromAgent, to: event.toAgent },
            };
            break;

          case 'finish':
            yield { event: 'message_done', data: { ...base, finish_reason: event.finishReason } };
            break;
        }
      }
    } catch (error) {
      hasError = true;
      yield {
        event: 'error',
        data: { session_id: sessionId, error: error instanceof Error ? error.message : String(error), code: 'LLM_ERROR' },
      };
    }

    if (fullContent) {
      const msg = this.messageRepo.create({
        id: messageId,
        role: 'assistant' as MessageRole,
        content: fullContent,
        session: { id: sessionId } as SessionEntity,
      });
      await this.messageRepo.save(msg);

      await this.sessionRepo.update(sessionId, { updated_at: new Date() });

      // 异步生成对话标题（仅当标题为默认值时）
      this.generateTitleIfNeeded(sessionId, messages, fullContent).catch(() => {});
    }

    if (!hasError) {
      yield { event: 'done', data: {} };
    }
  }

  async createSession(dto: CreateSessionDto): Promise<SessionEntity> {
    const session = this.sessionRepo.create({
      id: uuidv4(),
      title: dto.title || 'New Chat',
      is_pinned: false,
    });
    return this.sessionRepo.save(session);
  }

  async getSessions(page = 1, pageSize = 20) {
    const [all, total] = await this.sessionRepo.findAndCount({
      relations: ['messages'],
      order: { is_pinned: 'DESC', updated_at: 'DESC' },
      skip: (page - 1) * pageSize,
      take: pageSize,
    });

    const totalPages = Math.ceil(total / pageSize);
    const sessions = all.map(({ messages, ...rest }) => ({ ...rest, message_count: messages.length }));

    return {
      data: sessions,
      pagination: { total, page, page_size: pageSize, total_pages: totalPages },
    };
  }

  async getSession(id: string): Promise<SessionEntity> {
    const session = await this.sessionRepo.findOne({
      where: { id },
      relations: ['messages'],
    });
    if (!session) throw new NotFoundException(`Session ${id} not found`);
    return session;
  }

  async togglePin(id: string, dto: UpdateSessionDto): Promise<SessionEntity> {
    const session = await this.getSession(id);
    session.is_pinned = dto.is_pinned;
    session.updated_at = new Date();
    return this.sessionRepo.save(session);
  }

  async getMessages(sessionId: string): Promise<MessageEntity[]> {
    const session = await this.getSession(sessionId);
    return session.messages;
  }

  async deleteSession(id: string): Promise<void> {
    const session = await this.getSession(id);
    await this.messageRepo.delete({ session: { id: session.id } } as any);
    await this.sessionRepo.delete(session.id);
  }

  private async buildSystemPrompt(agentId?: string, ruleIds?: string[], webSearch?: boolean): Promise<string> {
    const parts: string[] = [];

    const resolvedAgentId = agentId || 'builtin-general';
    try {
      const agent = await this.agentService.findOne(resolvedAgentId);
      if (agent.system_prompt) parts.push(agent.system_prompt);
    } catch (e) {
      console.warn(`[ChatService] Failed to load agent ${resolvedAgentId}:`, e instanceof Error ? e.message : e);
    }

    // 自动注入当前时间，让 Agent 始终感知当前日期
    const now = new Date();
    const timeContext = `当前时间：${now.toLocaleDateString('zh-CN', { year: 'numeric', month: 'long', day: 'numeric', weekday: 'long' })} ${now.toLocaleTimeString('zh-CN', { hour: '2-digit', minute: '2-digit' })}`;
    parts.push(timeContext);

    if (webSearch) {
      parts.push(
        '[联网搜索已开启] 请先判断用户意图：如果用户询问最新信息、新闻、实时数据、近期事件等需要最新数据的请求，调用 web_search tool 搜索相关信息后基于搜索结果回答；对于普通对话、知识问答、代码编写等不需要实时数据的请求，不要调用 web_search tool。',
      );
    } else {
      parts.push(
        '[联网搜索已关闭] 不要调用 web_search tool。直接用已有知识回答用户问题。',
      );
    }

    // 1. 全局规则：强制注入
    const globalRules = await this.ruleService.getGlobalRules();

    // 2. Agent 级 general 规则：优先用前端传的 rule_ids，否则用 Agent 实体的 rule_ids
    let agentRuleIds = ruleIds;
    if (!agentRuleIds || agentRuleIds.length === 0) {
      try {
        const agent = await this.agentService.findOne(resolvedAgentId);
        agentRuleIds = agent.rule_ids || [];
      } catch {
        agentRuleIds = [];
      }
    }
    const agentRules = await this.ruleService.getRulesByIds(agentRuleIds);

    // 合并去重（global 优先级高）
    const allRules = [...globalRules];
    const globalIds = new Set(globalRules.map((r) => r.id));
    for (const r of agentRules) {
      if (!globalIds.has(r.id)) allRules.push(r);
    }

    if (allRules.length > 0) {
      parts.push(...allRules.map((r) => r.content));
    }

    const memoryContext = await this.memoryService.buildMemoryContext();
    if (memoryContext) parts.push(memoryContext);

    return parts.join('\n\n');
  }

  /**
   * 异步生成对话标题（仅当标题为默认值 "New Chat" 时触发）
   * 使用轻量 LLM 调用，不阻塞主流程
   */
  private async generateTitleIfNeeded(
    sessionId: string,
    historyMessages: { role: string; content: string }[],
    assistantReply: string,
  ): Promise<void> {
    const session = await this.sessionRepo.findOne({ where: { id: sessionId } });
    if (!session || session.title !== 'New Chat') return;

    // 取用户第一条消息和 AI 回复的前 200 字符
    const firstUserMsg = historyMessages.find((m) => m.role === 'user');
    if (!firstUserMsg) return;

    const userContent = firstUserMsg.content.substring(0, 200);
    const aiContent = assistantReply.substring(0, 200);

    try {
      const model = new ChatOpenAI({
        modelName: this.configService.openaiModel,
        openAIApiKey: this.configService.openaiApiKey,
        configuration: { baseURL: this.configService.openaiBaseUrl || undefined },
        maxTokens: 30,
        temperature: 0,
      });

      const result = await model.invoke([
        new SystemMessage(
          '根据以下对话内容生成一个简洁的标题（不超过15个中文字符）。要求：直接输出标题，不要任何解释、标点、引号。突出核心话题或意图。标题语言与对话语言一致。',
        ),
        new HumanMessage(`用户：${userContent}\nAI：${aiContent}`),
      ]);

      const title = (result.content as string).trim().replace(/[""''""。！？，、]/g, '');
      if (title && title.length > 0 && title.length <= 30) {
        await this.sessionRepo.update(sessionId, { title });
        console.log(`[ChatService] 标题已生成: "${title}" (session: ${sessionId.substring(0, 8)}...)`);
      }
    } catch (e) {
      // 标题生成失败不影响主流程，静默处理
      console.warn('[ChatService] 标题生成失败:', e instanceof Error ? e.message : e);
    }
  }
}
